File: examples/companion-packages/gem-lib/lib/tiny_flags.rb
# frozen_string_literal: true

require "digest"

# TinyFlags is a minimal feature-flag library with deterministic rollout bucketing.
#
# Two flag kinds:
#   - BooleanFlag(default:, rollout: nil) — on/off with optional percentage rollout.
#   - VariantFlag(variants:, default:, rollout: nil) — named treatment with optional allocations.
#
# Evaluation context is a Hash with optional :user_id, :account_id, or :session_id keys.
# Bucketing uses Digest::SHA1 over "<flag_name>:<context_id>" → integer in 0..99.
module TinyFlags
  # Validate that a percentage is an Integer in 0..100. Booleans are rejected.
  def self.validate_percentage!(value, label)
    if value.is_a?(TrueClass) || value.is_a?(FalseClass) || !value.is_a?(Integer)
      raise TypeError, "#{label} must be an Integer from 0 to 100"
    end
    raise ArgumentError, "#{label} must be an Integer from 0 to 100" if value.negative? || value > 100

    value
  end

  # Boolean feature flag with optional percentage rollout.
  class BooleanFlag
    attr_reader :default, :rollout

    def initialize(default: false, rollout: nil)
      unless default.is_a?(TrueClass) || default.is_a?(FalseClass)
        raise TypeError, "BooleanFlag default must be true or false"
      end

      @default = default
      @rollout = rollout.nil? ? nil : TinyFlags.validate_percentage!(rollout, "BooleanFlag rollout")
      freeze
    end
  end

  # Named-variant flag with optional per-variant allocations.
  class VariantFlag
    attr_reader :variants, :default, :rollout

    def initialize(variants:, default:, rollout: nil)
      unless variants.is_a?(Array) && !variants.empty?
        raise ArgumentError, "VariantFlag requires at least one variant"
      end

      string_variants = variants.map(&:to_s)
      if string_variants.uniq.length != string_variants.length || string_variants.any?(&:empty?)
        raise ArgumentError, "VariantFlag variants must be unique non-empty strings"
      end

      default_str = default.to_s
      raise ArgumentError, "VariantFlag default must be one of the variants" unless string_variants.include?(default_str)

      @variants = string_variants.freeze
      @default = default_str

      if rollout.nil?
        @rollout = nil
      else
        raise ArgumentError, "VariantFlag rollout must be a Hash" unless rollout.is_a?(Hash)

        total = 0
        normalized = {}
        rollout.each do |name, percentage|
          name_str = name.to_s
          unless string_variants.include?(name_str)
            raise ArgumentError, "VariantFlag rollout references unknown variant: #{name_str}"
          end

          normalized[name_str] = TinyFlags.validate_percentage!(percentage, "rollout for #{name_str.inspect}")
          total += normalized[name_str]
        end
        raise ArgumentError, "VariantFlag rollout percentages cannot exceed 100" if total > 100

        @rollout = normalized.freeze
      end

      freeze
    end
  end

  # A frozen set of flag definitions evaluated against a Hash context.
  class Registry
    def initialize(definitions)
      raise TypeError, "TinyFlags::Registry requires a definitions Hash" unless definitions.is_a?(Hash)

      @definitions = {}
      definitions.each do |name, flag|
        unless flag.is_a?(BooleanFlag) || flag.is_a?(VariantFlag)
          raise TypeError, "Definition for #{name.inspect} must be BooleanFlag or VariantFlag"
        end

        @definitions[name.to_s] = flag
      end
      @definitions.freeze
      freeze
    end

    attr_reader :definitions

    def names
      @definitions.keys
    end

    def include?(name)
      @definitions.key?(name.to_s)
    end

    def enabled?(name, context = nil)
      flag = lookup(name)
      raise TypeError, "TinyFlags flag #{name.inspect} is not a boolean flag" unless flag.is_a?(BooleanFlag)
      return flag.default if flag.rollout.nil?

      TinyFlags.bucket(name.to_s, context) < flag.rollout
    end

    def variant(name, context = nil)
      flag = lookup(name)
      raise TypeError, "TinyFlags flag #{name.inspect} is not a variant flag" unless flag.is_a?(VariantFlag)
      return flag.default if flag.rollout.nil?

      bucket = TinyFlags.bucket(name.to_s, context)
      upper_bound = 0
      flag.rollout.each do |variant_name, percentage|
        upper_bound += percentage
        return variant_name if bucket < upper_bound
      end
      flag.default
    end

    def evaluate(name, context = nil)
      flag = lookup(name)
      flag.is_a?(BooleanFlag) ? enabled?(name, context) : variant(name, context)
    end

    private

    def lookup(name)
      key = name.to_s
      flag = @definitions[key]
      raise KeyError, "Unknown TinyFlags flag: #{key}" if flag.nil?

      flag
    end
  end

  # Compute the deterministic 0..99 bucket for a flag name and context.
  # Context may be nil or a Hash with :user_id / :account_id / :session_id.
  def self.bucket(flag_name, context)
    key = "anonymous"
    if context.is_a?(Hash)
      key = context[:user_id] || context[:account_id] || context[:session_id] ||
            context["user_id"] || context["account_id"] || context["session_id"] ||
            "anonymous"
    end
    digest = Digest::SHA1.hexdigest("#{flag_name}:#{key}")
    digest[0, 8].to_i(16) % 100
  end
end

File: examples/companion-packages/lua-luarocks-lib/src/tinyflags.lua
--- TinyFlags is a minimal feature-flag library with deterministic rollout bucketing.
--
-- Two flag kinds:
--   * BooleanFlag{ default = bool, rollout = int? }   on/off with optional percentage rollout.
--   * VariantFlag{ variants = {...}, default = str, rollout = { name = int, ... }? }
--     named treatment with optional per-variant allocations.
--
-- Evaluation context is a table with optional `user_id`, `account_id`, or `session_id` keys.
-- Bucketing uses a pure-Lua FNV-1a 32-bit hash of "<flag_name>:<context_id>" mod 100.
--
-- The library is intentionally dependency-free so it runs on Lua 5.1+ and LuaJIT
-- without external hashing rocks.

local M = {}

local function is_integer(value)
  return type(value) == "number" and value == math.floor(value)
end

local function validate_percentage(value, label)
  if type(value) == "boolean" or not is_integer(value) then
    error(label .. " must be an integer from 0 to 100", 2)
  end
  if value < 0 or value > 100 then
    error(label .. " must be an integer from 0 to 100", 2)
  end
  return value
end

-- FNV-1a 32-bit. Pure Lua, no external deps. Result fits in 32 bits.
local FNV_OFFSET = 2166136261
local FNV_PRIME = 16777619
local MASK_32 = 0xFFFFFFFF

local function fnv1a(str)
  local hash = FNV_OFFSET
  for i = 1, #str do
    hash = (hash ~ string.byte(str, i)) & MASK_32
    hash = (hash * FNV_PRIME) & MASK_32
  end
  return hash
end

--- Compute the deterministic 0..99 bucket for a flag name and context.
-- @param flag_name string
-- @param context  table | nil
function M.bucket(flag_name, context)
  local key = "anonymous"
  if type(context) == "table" then
    key = context.user_id or context.account_id or context.session_id or "anonymous"
  end
  return fnv1a(flag_name .. ":" .. tostring(key)) % 100
end

local BooleanFlag = {}
BooleanFlag.__index = BooleanFlag

--- Create a boolean feature flag.
-- @param opts table { default = bool, rollout = int? }
function M.BooleanFlag(opts)
  if type(opts) ~= "table" then
    error("BooleanFlag requires an options table", 2)
  end
  if type(opts.default) ~= "boolean" then
    error("BooleanFlag default must be true or false", 2)
  end
  local rollout = nil
  if opts.rollout ~= nil then
    rollout = validate_percentage(opts.rollout, "BooleanFlag rollout")
  end
  local flag = setmetatable({
    kind = "boolean",
    default = opts.default,
    rollout = rollout,
  }, BooleanFlag)
  return flag
end

local VariantFlag = {}
VariantFlag.__index = VariantFlag

local function copy_string_array(arr, label)
  if type(arr) ~= "table" or #arr == 0 then
    error(label .. " requires at least one variant", 2)
  end
  local out = {}
  local seen = {}
  for i = 1, #arr do
    local v = arr[i]
    if type(v) ~= "string" or v == "" then
      error(label .. " variants must be unique non-empty strings", 2)
    end
    if seen[v] then
      error(label .. " variants must be unique non-empty strings", 2)
    end
    seen[v] = true
    out[i] = v
  end
  return out, seen
end

--- Create a variant flag with allowed values and an optional rollout map.
-- @param opts table { variants = {...}, default = str, rollout = { name = int, ... }? }
function M.VariantFlag(opts)
  if type(opts) ~= "table" then
    error("VariantFlag requires an options table", 2)
  end
  local variants, seen = copy_string_array(opts.variants, "VariantFlag")
  if type(opts.default) ~= "string" or not seen[opts.default] then
    error("VariantFlag default must be one of the variants", 2)
  end

  local rollout = nil
  if opts.rollout ~= nil then
    if type(opts.rollout) ~= "table" then
      error("VariantFlag rollout must be a table", 2)
    end
    rollout = {}
    local total = 0
    for name, percentage in pairs(opts.rollout) do
      if not seen[name] then
        error("VariantFlag rollout references unknown variant: " .. tostring(name), 2)
      end
      rollout[name] = validate_percentage(percentage, "rollout for '" .. name .. "'")
      total = total + rollout[name]
    end
    if total > 100 then
      error("VariantFlag rollout percentages cannot exceed 100", 2)
    end
  end

  return setmetatable({
    kind = "variant",
    variants = variants,
    default = opts.default,
    rollout = rollout,
  }, VariantFlag)
end

local Registry = {}
Registry.__index = Registry

--- Create a registry from a definitions table.
-- @param definitions table { [name] = BooleanFlag | VariantFlag, ... }
function M.Registry(definitions)
  if type(definitions) ~= "table" then
    error("TinyFlags.Registry requires a definitions table", 2)
  end
  local defs = {}
  local order = {}
  for name, flag in pairs(definitions) do
    if type(flag) ~= "table" or (flag.kind ~= "boolean" and flag.kind ~= "variant") then
      error("Definition for '" .. tostring(name) .. "' must be BooleanFlag or VariantFlag", 2)
    end
    defs[name] = flag
    order[#order + 1] = name
  end
  return setmetatable({ definitions = defs, _order = order }, Registry)
end

local function lookup(self, name)
  local flag = self.definitions[name]
  if flag == nil then
    error("Unknown TinyFlags flag: " .. tostring(name), 3)
  end
  return flag
end

function Registry:names()
  local out = {}
  for i = 1, #self._order do
    out[i] = self._order[i]
  end
  return out
end

function Registry:has(name)
  return self.definitions[name] ~= nil
end

function Registry:is_enabled(name, context)
  local flag = lookup(self, name)
  if flag.kind ~= "boolean" then
    error("TinyFlags flag '" .. name .. "' is not a boolean flag", 2)
  end
  if flag.rollout == nil then
    return flag.default
  end
  return M.bucket(name, context) < flag.rollout
end

function Registry:variant(name, context)
  local flag = lookup(self, name)
  if flag.kind ~= "variant" then
    error("TinyFlags flag '" .. name .. "' is not a variant flag", 2)
  end
  if flag.rollout == nil then
    return flag.default
  end
  local bucket = M.bucket(name, context)
  local upper = 0
  -- Iterate in declared variant order for stable allocation.
  for i = 1, #flag.variants do
    local v = flag.variants[i]
    local pct = flag.rollout[v]
    if pct ~= nil then
      upper = upper + pct
      if bucket < upper then
        return v
      end
    end
  end
  return flag.default
end

function Registry:evaluate(name, context)
  local flag = lookup(self, name)
  if flag.kind == "boolean" then
    return self:is_enabled(name, context)
  end
  return self:variant(name, context)
end

return M
